File: resources/recipes/databases/optimistic-locking/optimistic_update.py
"""Optimistic locking example with PostgreSQL.

Run: pip install psycopg2-binary
Set: DATABASE_URL environment variable or conn string
"""
import os

import psycopg2
from psycopg2.extras import RealDictCursor


def get_connection():
    return psycopg2.connect(os.environ.get("DATABASE_URL", "postgresql://localhost/test"))


def update_product_price(conn, product_id: int, new_price: float, expected_version: int):
    with conn.cursor(cursor_factory=RealDictCursor) as cur:
        cur.execute(
            """
            UPDATE products
            SET price = %s, version = version + 1
            WHERE id = %s AND version = %s
            RETURNING id, version;
            """,
            (new_price, product_id, expected_version),
        )
        updated = cur.fetchone()
        if not updated:
            raise ValueError(
                f"Conflict: product {product_id} was modified by another transaction. "
                "Please refresh and retry."
            )
        conn.commit()
        return updated


def batch_update_prices(conn, updates):
    """Update multiple rows with optimistic locking in one transaction."""
    results = []
    with conn.cursor() as cur:
        for item in updates:
            cur.execute(
                """
                UPDATE products
                SET price = %s, version = version + 1
                WHERE id = %s AND version = %s
                RETURNING id, version;
                """,
                (item["new_price"], item["id"], item["expected_version"]),
            )
            updated = cur.fetchone()
            if not updated:
                conn.rollback()
                raise ValueError(
                    f"Conflict on product {item['id']}: "
                    f"expected version {item['expected_version']}"
                )
            results.append(updated)
    conn.commit()
    return results


if __name__ == "__main__":
    conn = get_connection()
    try:
        result = update_product_price(conn, product_id=1, new_price=19.99, expected_version=3)
        print(f"Updated to version {result['version']}")
    except ValueError as e:
        print(e)

File: resources/recipes/databases/optimistic-locking/optimistic_update.js
// Optimistic locking example with PostgreSQL and Node.js.
//
// Run: npm install pg
// Set: DATABASE_URL or PG* env vars

const { Pool } = require('pg');

const pool = new Pool({
  // Use env vars or pass connection config
});

async function updateProductPrice(productId, newPrice, expectedVersion) {
  const result = await pool.query(
    `UPDATE products
     SET price = $1, version = version + 1, updated_at = NOW()
     WHERE id = $2 AND version = $3
     RETURNING id, version;`,
    [newPrice, productId, expectedVersion]
  );

  if (result.rowCount === 0) {
    const current = await pool.query(
      'SELECT version FROM products WHERE id = $1',
      [productId]
    );
    throw new Error(
      `Version conflict: expected ${expectedVersion}, found ${current.rows[0]?.version}. Please retry.`
    );
  }

  return result.rows[0];
}

async function withRetry(fn, maxRetries = 3, baseDelay = 50) {
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      return await fn();
    } catch (err) {
      if (!err.message.includes('Version conflict') || attempt === maxRetries - 1) {
        throw err;
      }
      const delay = baseDelay * (2 ** attempt) + Math.random() * 50;
      await new Promise((r) => setTimeout(r, delay));
    }
  }
}

// Express route
const express = require('express');
const app = express();
app.use(express.json());

app.put('/products/:id', async (req, res) => {
  try {
    const product = await updateProductPrice(
      req.params.id,
      req.body.price,
      req.body.version
    );
    res.json(product);
  } catch (err) {
    res.status(409).json({ error: err.message });
  }
});

if (require.main === module) {
  app.listen(3000, () => console.log('Listening on :3000'));
}
